faucet/src/lib.rs

```diff
@@ -56,125 +63,277 @@ pub mod pallet {
 
+			if !<Faucets<T>>::contains_key(&faucet_account) {
+				return Err(<Error<T>>::InexsistentFaucet.into())
+			}
+
+			let faucet = Self::faucets(&faucet_account).ok_or(<Error<T>>::InexsistentFaucet)?;
+
+			if !faucet.whitelist.contains(&cid) {
+				return Err(<Error<T>>::CommunityNotInWhitelist.into())
+			}
+
 			<encointer_reputation_commitments::Pallet<T>>::do_commit_reputation(
 				&from,
 				cid,
 				cindex,
-				Self::reputation_commitments_purpose_id(),
+				faucet.purpose_id,
 				None,
 			)?;
 
-			T::Currency::transfer(&Self::account_id(), &from, Self::drip_amount(), KeepAlive)
-				.map_err(|_| <Error<T>>::FaucetEmpty)?;
+			<T as Config>::Currency::transfer(
+				&faucet_account,
+				&from,
+				faucet.drip_amount,
+				KeepAlive,
+			)
+			.map_err(|_| <Error<T>>::FaucetEmpty)?;
 
-			Self::deposit_event(Event::Dripped(from, Self::drip_amount()));
+			info!(
+				target: LOG,
+				"faucet {:?} dripped {:?} to {:?}", faucet.name, faucet.drip_amount, from
+			);
+			Self::deposit_event(Event::Dripped(faucet_account, from, faucet.drip_amount));
 
 			Ok(().into())
 		}
 
-		#[pallet::call_index(1)]
+		#[pallet::call_index(2)]
 		#[pallet::weight(10_000)]
-		pub fn set_drip_amount(
+		pub fn dissolve_faucet(
 			origin: OriginFor<T>,
-			drip_amount: BalanceOf<T>,
+			faucet_account: T::AccountId,
+			beneficiary: T::AccountId,
 		) -> DispatchResultWithPostInfo {
-			T::ControllerOrigin::ensure_origin(origin)?;
-			<DripAmount<T>>::put(drip_amount);
-			info!(target: LOG, "set drip amount to {:?}", drip_amount);
-			Self::deposit_event(Event::DripAmountUpdated(drip_amount));
+			ensure_root(origin)?;
+
+			let faucet = Self::faucets(&faucet_account).ok_or(<Error<T>>::InexsistentFaucet)?;
+
+			<T as Config>::Currency::unreserve_all_named(
+				&Self::get_reserve_id(&faucet_account),
+				&faucet.creator,
+			);
+
+			<Faucets<T>>::remove(&faucet_account);
+			<T as Config>::Currency::transfer(
+				&faucet_account,
+				&beneficiary,
+				<T as Config>::Currency::free_balance(&faucet_account),
+				AllowDeath,
+			)?;
+
+			info!(target: LOG, "faucet dissolved {:?}", faucet_account);
+			Self::deposit_event(Event::FaucetDissolved(faucet_account));
 			Ok(().into())
 		}
-	}
 
-	impl<T: Config> Pallet<T> {
-		/// The account ID of the faucet.
-		///
-		/// This actually does computation. If you need to keep using it, then make sure you cache the
-		/// value and only call this once.
-		pub fn account_id() -> T::AccountId {
-			T::PalletId::get().into_account_truncating()
+		#[pallet::call_index(3)]
+		#[pallet::weight(10_000)]
+		pub fn close_faucet(
+			origin: OriginFor<T>,
+			faucet_account: T::AccountId,
+		) -> DispatchResultWithPostInfo {
+			let from = ensure_signed(origin)?;
+			let faucet = Self::faucets(&faucet_account).ok_or(<Error<T>>::InexsistentFaucet)?;
+
+			ensure!(from == faucet.creator, <Error<T>>::NotCreator);
+			ensure!(
+				<T as Config>::Currency::free_balance(&faucet_account) < faucet.drip_amount,
```

Comment:
I'm afraid this could go south: If faucet is created with an integer multiple of drip amount, there will be a moment when faucet balance = drip_amount. But you can't drip anymore because of keep_alive...it wont go below ED
to be safe, the test could be "if balance < 2*drip_amount